src/db/commit.rs

```diff
@@ -277,10 +300,153 @@ pub enum FromHashError {
+            &store.read().await.unwrap().read(),
+        )
+        .await
+        .unwrap();
+        chain.push(commit);
+        chain
+    }
+
+    pub async fn add_local<'a>(chain: &'a mut Chain, store: &dag::Store) -> &'a mut Chain {
+        assert!(chain.len() > 0);
+        let w = Write::new_local(
+            Whence::Head(str!("main")),
+            str!("mutator_name"),
+            Any::Array(vec![]),
+            None,
+            store.write().await.unwrap(),
+        )
+        .await
+        .unwrap();
+        w.commit("main", "local_create_date", "checksum")
+            .await
+            .unwrap();
+        let (_, commit, _) = read_commit(
+            Whence::Head(str!("main")),
+            &store.read().await.unwrap().read(),
+        )
+        .await
+        .unwrap();
+        chain.push(commit);
+        chain
+    }
+
+    pub async fn add_snapshot<'a>(chain: &'a mut Chain, store: &dag::Store) -> &'a mut Chain {
+        assert!(chain.len() > 0);
+        let w = Write::new_snapshot(
+            Whence::Head(str!("main")),
+            chain[chain.len() - 1].next_mutation_id() - 1,
+            str!("server_state_id"),
+            store.write().await.unwrap(),
+        )
+        .await
+        .unwrap();
+        w.commit("main", "local_create_date", "checksum")
+            .await
+            .unwrap();
+        let (_, commit, _) = read_commit(
+            Whence::Head(str!("main")),
+            &store.read().await.unwrap().read(),
+        )
+        .await
+        .unwrap();
+        chain.push(commit);
+        chain
+    }
+}
+
 #[cfg(test)]
 mod tests {
+    use super::test_helpers::*;
     use super::*;
     use crate::dag::Chunk;
+    use crate::kv::memstore::MemStore;
+
+    #[async_std::test]
+    async fn test_base_snapshot() {
+        let store = dag::Store::new(Box::new(MemStore::new()));
+        let mut chain: Chain = vec![];
+
+        add_genesis(&mut chain, &store).await;
+        let genesis_hash = chain[0].chunk().hash();
+        assert_eq!(
+            genesis_hash,
+            Commit::base_snapshot(genesis_hash, &store.read().await.unwrap().read())
+                .await
+                .unwrap()
+                .chunk()
+                .hash()
+        );
+
+        add_local(&mut chain, &store).await;
+        add_local(&mut chain, &store).await;
+        let genesis_hash = chain[0].chunk().hash();
+        assert_eq!(
+            genesis_hash,
+            Commit::base_snapshot(
+                chain[chain.len() - 1].chunk().hash(),
+                &store.read().await.unwrap().read()
+            )
+            .await
+            .unwrap()
+            .chunk()
+            .hash()
+        );
+
+        add_snapshot(&mut chain, &store).await;
+        let base_hash = store
+            .read()
+            .await
+            .unwrap()
+            .read()
+            .get_head("main")
+            .await
+            .unwrap()
+            .unwrap();
```

Comment:
this deserves some kind of award.

Reply:
do you think we should adjust our fmt parameters to allow longer lines? you can barely follow the flow of the function when a single let statement is spread over ten vertical lines.